src/wpt/harness.ts

```diff
@@ -66,8 +68,208 @@ type TestCase = {
+  COMPLETE = 4,
+}
+
+/**
+ * @class
+ *
+ * A single subtest. A Test is not constructed directly but via the
+ * :js:func:`test`, :js:func:`async_test` or :js:func:`promise_test` functions.
+ *
+ * @param {string} name - This must be unique in a given file and must be
+ * invariant between runs.
+ *
+ */
+class Test {
+  public name: string;
+  public properties: unknown;
+  public phase: TestPhase;
+
+  public error?: Error;
+
+  // For convenience, expose a promise that resolves once done() is called
+  public isDone: Promise<void>;
+  private resolve: () => void;
+
+  public constructor(name: string, properties: unknown) {
+    this.name = name;
+    this.properties = properties;
+    this.phase = TestPhase.INITIAL;
+
+    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type -- void is being used as a valid generic in this context
+    const { promise, resolve } = Promise.withResolvers<void>();
+    this.isDone = promise;
+    this.resolve = resolve;
+  }
+
+  /**
+   * Run a single step of an ongoing test.
+   *
+   * @param {string} func - Callback function to run as a step. If
+   * this throws an :js:func:`AssertionError`, or any other
+   * exception, the :js:class:`Test` status is set to ``FAIL``.
+   * @param {Object} [this_obj] - The object to use as the this
+   * value when calling ``func``. Defaults to the  :js:class:`Test` object.
+   */
+  public step(
+    func: UnknownFunc,
+    this_obj?: object,
+    ...rest: unknown[]
+  ): unknown {
+    if (this.phase > TestPhase.STARTED) {
+      return undefined;
+    }
+
+    if (arguments.length === 1) {
+      // eslint-disable-next-line @typescript-eslint/no-this-alias -- WPT allows for overriding the this environment for a step but defaults to the Test class
+      this_obj = this;
+    }
+
+    try {
+      return func.call(this_obj, ...rest);
+    } catch (err) {
+      if (this.phase >= TestPhase.HAS_RESULT) {
+        return undefined;
+      }
+
+      this.error = new AggregateError([err], this.name);
+      this.done();
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Wrap a function so that it runs as a step of the current test.
+   *
+   * This allows creating a callback function that will run as a
+   * test step.
+   *
+   * @example
+   * let t = async_test("Example");
+   * onload = t.step_func(e => {
+   *   assert_equals(e.name, "load");
+   *   // Mark the test as complete.
+   *   t.done();
+   * })
+   *
+   * @param {string} func - Function to run as a step. If this
+   * throws an :js:func:`AssertionError`, or any other exception,
+   * the :js:class:`Test` status is set to ``FAIL``.
+   * @param {Object} [this_obj] - The object to use as the this
+   * value when calling ``func``. Defaults to the :js:class:`Test` object.
+   */
+  public step_func(func: UnknownFunc, this_obj?: object): UnknownFunc {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias -- We could get rid of this this alias but the code is copied from WPT
```

Comment:
Can you add this to the beginning of this file, so we don't need to disable everywhere